pymc/step_methods/state.py

```diff
@@ -0,0 +1,98 @@
+#   Copyright 2024 The PyMC Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+from copy import deepcopy
+from dataclasses import dataclass, fields
+from typing import Any
+
+import numpy as np
+
+
+class MetaDataClassState(type):
+    def __new__(cls, name, bases, namespace):
+        return dataclass(eq=False, kw_only=True)(super().__new__(cls, name, bases, namespace))
+
+
+class DataClassState(metaclass=MetaDataClassState):
+    def __eq__(self, other: Any) -> bool:
+        if self.__class__ != other.__class__:
+            return False
+        this_fields = set([f.name for f in fields(self)])
+        other_fields = set([f.name for f in fields(other)])
```

Comment:
inner list comprehension is unnecessary

Reply:
Why not? The `fields` function will return `Field` objects, that have a bunch of extra dataclass specific attributes (e.g. `type`, `default`, `default_factory`). I just want to check that the names are the same and use those names later for `getattr`.

Reply:
Oh! I think I understand what you're saying. A `set` of `Field` objects should already be enough to test if `this_fields == other_fields` because all of the other attributes should also match.

Reply:
actually that was not my point (but you might be right about it)

`set(generator)` aka `set(a for a in "ABCD")` works. You can leave out creating the inner list (and then iterating it again when creating the set)